enterprise/cmd/frontend/internal/auth/gitlaboauth/session_test.go

```diff
@@ -3,26 +3,198 @@ package gitlaboauth
+				glUser: &gitlab.User{
+					ID:       int32(101),
+					Username: string("alice"),
+					Email:    string("alice@example.com"),
+				},
+				glUserGroups: []*gitlab.Group{
+					{FullPath: "group1"},
+					{FullPath: "group2"},
+				},
+				glUserGroupsErr: errors.New("boom"),
+			}},
+			expErr: true,
+		},
+		{
+			inputs: []input{{
+				description: "glUser, in allowed groups -> session created",
+				allowGroups: []string{"group1"},
+				glUser: &gitlab.User{
+					ID:       int32(101),
+					Username: string("alice"),
+					Email:    string("alice@example.com"),
+				},
+				glUserGroups: []*gitlab.Group{
+					{FullPath: "group1"},
+					{FullPath: "group2"},
+				},
+			}},
+			expActor: &actor.Actor{UID: 1},
+			expAuthUserOp: &auth.GetAndSaveUserOp{
+				UserProps:        u("alice", "alice@example.com", true),
+				ExternalAccount:  acct(extsvc.TypeGitLab, "https://gitlab.com/", clientID, "101"),
+				CreateIfNotExist: true,
+			},
+		},
+	}
+
+	for _, c := range cases {
+		for _, ci := range c.inputs {
+			c, ci := c, ci
+
+			t.Run(ci.description, func(t *testing.T) {
+
+				gitlab.MockListGroups = func(ctx context.Context, page int) (groups []*gitlab.Group, hasNextPage bool, err error) {
+					return ci.glUserGroups, false, ci.glUserGroupsErr
+				}
+
+				var gotAuthUserOp *auth.GetAndSaveUserOp
+				auth.MockGetAndSaveUser = func(ctx context.Context, op auth.GetAndSaveUserOp) (userID int32, safeErrMsg string, err error) {
+					if gotAuthUserOp != nil {
+						t.Fatal("GetAndSaveUser called more than once")
+					}
+
+					op.ExternalAccountData = extsvc.AccountData{}
+					gotAuthUserOp = &op
+
+					if uid, ok := authSaveableUsers[op.UserProps.Username]; ok {
+						return uid, "", nil
+					}
+
+					return 0, "safeErr", errors.New("error mocking get and save user")
+				}
+
+				defer func() {
+					auth.MockGetAndSaveUser = nil
+				}()
+
+				ctx := WithUser(context.Background(), ci.glUser)
+				s := &sessionIssuerHelper{
+					CodeHost:    codeHost,
+					clientID:    clientID,
+					allowGroups: ci.allowGroups,
+				}
+
+				tok := &oauth2.Token{AccessToken: "dummy-value-that-isnt-relevant-to-unit-correctness"}
+				actr, _, err := s.GetOrCreateUser(ctx, tok, "", "", "")
+
+				if got, exp := actr, c.expActor; !reflect.DeepEqual(got, exp) {
+					t.Errorf("expected actor %v, got %v", exp, got)
+				}
+
+				if c.expErr && err == nil {
+					t.Errorf("expected err %v, but was nil", c.expErr)
+				} else if !c.expErr && err != nil {
+					t.Errorf("expected no error, but was %v", err)
+				}
+
+				if got, exp := gotAuthUserOp, c.expAuthUserOp; !reflect.DeepEqual(got, exp) {
+					t.Error(cmp.Diff(got, exp))
+				}
+			})
+		}
+	}
+
+}
+
+func u(username, email string, emailIsVerified bool) database.NewUser {
+	return database.NewUser{
+		Username:        username,
+		Email:           email,
+		EmailIsVerified: emailIsVerified,
+	}
+}
+
+func acct(serviceType, serviceID, clientID, accountID string) extsvc.AccountSpec {
```

Comment:
I think we are not getting much benefit for having `u` vs `newUser`, besides losing quite a lot readability.

```suggestion
func newUser(username, email string, emailIsVerified bool) database.NewUser {
	return database.NewUser{
		Username:        username,
		Email:           email,
		EmailIsVerified: emailIsVerified,
	}
}

func newAccountSpec(serviceType, serviceID, clientID, accountID string) extsvc.AccountSpec {
```

Reply:
On second thought... I personally don't think these two functions have a reason to exist at all, they are literally composing structs without any extra logic.